src/wallet/wallet.cpp

```diff
@@ -4213,6 +4215,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();
 
+    // Register with the validation interface. Skip requesting mempool transactions if wallet is empty.
```

Comment:
I think moving the sync-from-mempool function to above the initial rescan breaks the (undocumented) expectation that the wallet is notified of transactions in dependency order. In fact, this was slightly broken by #15652, but this PR makes it worse.

Prior to dynamically-loaded wallets, the wallet would always receive a transaction after it had received all of that transaction's ancestors. That wallet relies on that expectation:

- `CWallet::IsFromMe()` will only match on a transaction if its parent is already in the wallet.
- `nTimeSmart` should be monotonically increasing for transaction descendency. If transactions arrive in the wrong order then this is no longer true, and they'll be shown out-of-order in the GUI.

Since #10740, a wallet could receive mempool transactions out-of-order (since a newly loaded wallet would not be notified of transactions in the mempool, but could be notified of their descendant transactions when they arrive).  #15652 improved the situtation by notifying the wallet of all mempool transactions, but didn't sort them first, so a wallet could still be notified of the mempool txs in the wrong order. This PR potentially makes things worse by notifying the wallet of mempool txs before the blocks which could contain the mempool txs' ancestors.

I think to fix this, we need to:
- move the sync-from-mempool to after the rescan call, while cs_main is still held
- sort the mempool txs by using `mempool.infoAll()` before notifying the wallet.

We should also document that the wallet expects to see transactions in order!

Reply:
I also wonder if there could brokenness in the case where a reorg happens during a `rescanblockchain` or `importmulti` RPC call (these rescans are different because cs_main isn't held the whole time).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r274934195

> I think to fix this, we need to:
> 
> * move the sync-from-mempool to after the rescan call, while cs_main is still held
> * sort the mempool txs by using `mempool.infoAll()` before notifying the wallet.
> 
> We should also document that the wallet expects to see transactions in order!

Moving sync is done. Rescanning happens before syncing from mempool in the current version of this PR. Sorting mempool isn't done yet. So this will still be as broken as it has been since https://github.com/bitcoin/bitcoin/pull/15652, but not more broken. There's already a lot going on here, so I think I would want to address this in a different PR. It seems pretty easy to fix independently of this PR, but writing a test case could be a challenge